### PR TITLE
Utilize pointer cache in default input system

### PR DESCRIPTION
**Pull request by Troy-Ferrell**
## Overview
This change introduces the concept of a pointer cache so that pointers, and in particular their prefabs, can be re-used throughout the life cycle of an app. The benefit of a cache is greatly improved performance by utilizing resource re-use. Instantiating prefabs/gameobjects in Unity is very expensive and must occur on the main thread. Particularly for HL2, where hands are lost/found every time a users raises or lowers them, the performance is critically impacted. 

This new cache is primarily implemented by extending the functionality of BaseInputDeviceManager. Currently the class supports requesting pointers based on some arguments for internal use. This adds the ability for the class to also recycle or clean up pointers as well. Furthermore, this cycle is extended to work with an object pool paradigm. If a pointer has not been destroyed (either by itself or some other system in the app), then the BaseInputDeviceManager will attempt to cache the pointer for later re-use. 

The base pointer implementation class, ControllerPoseSynchronizer.cs, has the property DestroyOnSourceLost which will allow the pointer to destroy itself if it's input source has been lost. In order for the caching to work, any pointer prefabs/instances need to disable this toggle.

Also adds `UnityObjectExtensions.IsNull()` utility helper to combat issue seen in #5981

## Changes
- Fixes: #5846 , #4223 , #5981

### CODE REVIEW

File: Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs [hunk 1]
```diff
@@ -61,36 +155,97 @@ public abstract class BaseInputDeviceManager : BaseDataProvider<IMixedRealityInp
...
+        {
+            if (inputSource != null)
+            {
+                var pointers = inputSource.Pointers;
+                for (int i = 0; i < pointers.Length; i++)
+                {
+                    var p = pointers[i];
+                    if (p != null && p is MonoBehaviour pointerComponent)
+                    {
+                        // TODO: Troy - look at reset method or other properties?
```
Comment (wiwei) [inline Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs]: >// TODO: Troy - look at reset method or other properties? [](start = 24, length = 57)

+1 It feels like there are other properties here we might want to reset (i.e. InputSourceParent, pointerName) - or maybe not? I think those will get reset to correct things once the controller is re-set the next time around, though it feels like a pointer going back into the cache should be as-close-as-possible to what happens after CreatePointer.

File: Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs [hunk 2]
```diff
@@ -61,36 +155,97 @@ public abstract class BaseInputDeviceManager : BaseDataProvider<IMixedRealityInp
...
+                    var pointerCache = pointerConfigurations[i].cache;
+                    if (pointerCache.Count > 0)
                     {
-                        var pointerObject = Object.Instantiate(pointerProfile.PointerPrefab);
-                        MixedRealityPlayspace.AddChild(pointerObject.transform);
-                        var pointer = pointerObject.GetComponent<IMixedRealityPointer>();
-
-                        if (pointer != null)
+                        var p = pointerCache.Pop();
+                        if (p != null && p is MonoBehaviour pointerGameObject)
```
Comment (wiwei) [inline Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs]: >if (p != null && p is MonoBehaviour pointerGameObject) [](start = 24, length = 54)

This is probably not the type of feedback being requested, but I did wanna leave a little comment about this one (i.e. it feels weird that we'd store stuff in the cache that we wouldn't consider later - I guess this is primarily to defend against cases where the game object is destroyed out from under us through other mechanisms right?)

Comment (Troy-Ferrell) [reply Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs]: @wiwei no happy for the feedback. I just didn't want nit stuff atm like forgot to add comment etc. 

So I agree, but I'm not sure how to overcome this. The pointer is a GameObject and anyone with a reference to it could at some pointer potentially destroy it. So it shouldn't ever be null but we have to check against it.

File: Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs [hunk 3]
```diff
@@ -53,6 +53,100 @@ public abstract class BaseInputDeviceManager : BaseDataProvider<IMixedRealityInp
...
+                {
+                    return pointer.GetHashCode();
+                }
+            }
+        }
+
+        // Active pointers associated with the config index they were spawned from
+        private Dictionary<IMixedRealityPointer, uint> activePointersToConfig 
+            = new Dictionary<IMixedRealityPointer, uint>(PointerEqualityComparer.Default);
+
```
Comment (wiwei) [inline Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs]: Another point to consider when doing this sort of change - I assume that we're gonna have bugs (because there are always bugs, it's a fact of life!)

It would be interesting to have a boolean/flag/control mechanism (maybe on this class, maybe on the profile itself) that will either turn on or turn off this behavior. This way we can really easily test if/when there are issues, if they are caused by recycling of pointers. i.e. if someone notices a bug, they can try turning off this new feature and see if the issue still repros, and if not, we have a better idea of what to look at.

Obviously such a flag would be removed soonish after in a future release, but in the interim these control mechanisms are really helpful for issue/feature isolation.

File: Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs [hunk 4]
```diff
@@ -61,36 +155,97 @@ public abstract class BaseInputDeviceManager : BaseDataProvider<IMixedRealityInp
         /// <param name="useSpecificType">Only register pointers with a specific type.</param>
         protected virtual IMixedRealityPointer[] RequestPointers(SupportedControllerType controllerType, Handedness controllingHand)
```
Comment (wiwei) [inline Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs]: >RequestPointers [](start = 49, length = 15)

We haven't really done this (at all) for other parts, but since we're doing perf work in this case, consider adding something like:

[https://docs.unity3d.com/ScriptReference/Profiling.Profiler.BeginSample.html](https://docs.unity3d.com/ScriptReference/Profiling.Profiler.BeginSample.html)

So that we can measure the impact (in the Unity profiler) or having pointer caching vs not caching.

File: Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs [hunk 5]
```diff
@@ -53,6 +53,100 @@ public abstract class BaseInputDeviceManager : BaseDataProvider<IMixedRealityInp
...
 
+        private struct PointerConfig
+        {
+            public PointerOption profile;
+            public Stack<IMixedRealityPointer> cache;
+        }
+
+        private PointerConfig[] pointerConfigurations = System.Array.Empty<PointerConfig>();
+
+        private class PointerEqualityComparer : IEqualityComparer<IMixedRealityPointer>
```
Comment (wiwei) [inline Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs]: > private class PointerEqualityComparer : IEqualityComparer<IMixedRealityPointer> [](start = 7, length = 80)

One other thing to think about here (overall structure - I know that this is also another comment you're not interested in) - around how this could be written to be a bit more testable - I assume that some of these could be torn out into a different class which can be individually unit/component tested, which MIGHT help.

I say might because I haven't actually done the deep thinking to know if it's actually valuable/not valuable, as it's possible that this level of testing isn't valuable, but it could be!

**Comment by wiwei** [comment]
Change generally seems sane to me (i.e. overall intent and how things are generally being done). I think that given the different set of things being stored on the pointer itself, there's definitely some room to figure out which of those need to also be reset (i.e. should we have some test/assert that the pointer that is generated clean should have the things set/unset compared to a cached one)

File: Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs (hunk 5 above)
Comment (Troy-Ferrell) [reply Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs]: > I assume that some of these could be torn out into a different class which can be individually unit/component tested, which MIGHT help.

@wiwei  I'm not exactly understanding you. What are "these" that could be torn out?

File: Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs (hunk 1 above)
Comment (julenka) [reply Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs]: Pointers may also want to reset visual state, is pressed state. Perhaps a reset() method on the IMRPointer interface which would reset to default state in case pointer gets lost, etc.

**Comment by julenka** [comment]
Requesting changes to reflect that I think adding tests to protect our future selves from  not resetting pointers correctly is important.

File: Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/GGVPointer.cs [hunk 6]
```diff
@@ -172,6 +172,7 @@ public override int GetHashCode()
             }
         }
 
+        /// <inheritdoc />
```
Comment (julenka) [inline Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/GGVPointer.cs]: Thanks for adding these doc -- "leave the code cleaner than when you arrived"

File: Assets/MixedRealityToolkit/Interfaces/InputSystem/IMixedRealityPointer.cs [hunk 7]
```diff
@@ -125,5 +125,10 @@ public interface IMixedRealityPointer : IEqualityComparer
         /// Called during the scene query just before the current pointer target changes.
         /// </summary>
         void OnPreCurrentPointerTargetChange();
+
+        /// <summary>
+        /// Resets pointer to initial state
```
Comment (julenka) [inline Assets/MixedRealityToolkit/Interfaces/InputSystem/IMixedRealityPointer.cs]: It would be good to add a "why" behind this so people understand the context -- we recycle pointers, not destroy them so when a hand is lost this should really reset the state completely.

File: Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs [hunk 8]
```diff
@@ -12,6 +13,13 @@ namespace Microsoft.MixedReality.Toolkit.Input
     /// </summary>
     public abstract class BaseInputDeviceManager : BaseDataProvider<IMixedRealityInputSystem>, IMixedRealityInputDeviceManager
     {
+        /// <summary>
+        /// Temporary control mechanism to enable/disable use of Pointer Cache in request/recycling of pointers by Input System
```
Comment (julenka) [inline Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs]: Temporary? Do you mean to remove this during this PR, or after this PR?

**Comment by Troy-Ferrell** [comment]
@julenka very good points. I think at least the basic test flow at a minimum should be

- Raise hand
- Click on objects
- Hide hand (source lost)
- Raise hand
- Click on object at different location
- Hide hand

Another potential one (which may already exist...I need to look at pointer events)
- Raise both hands and focus on same gameobject
- alternate show/hide of hands (one always remain in focus on button)
- then hand ray select button

File: Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs (hunk 8 above)
Comment (Troy-Ferrell) [reply Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs]: After this PR....I wish there was a #IF MRTK_DEV lol

I'll consider this more. At a min, I'll open a GitHub issue to track it's removal

**Comment by Troy-Ferrell** [comment]
@thalbern / @julenka added tests and cleaned up code

File: Assets/MixedRealityToolkit.SDK/Features/UX/Prefabs/Pointers/ParabolicPointer.prefab [hunk 9]
```diff
@@ -365,6 +365,9 @@ MonoBehaviour:
   lineMaterial: {fileID: 2100000, guid: 11727442de02c1d4b8d37a063c748aec, type: 2}
   roundedEdges: 1
   roundedCaps: 1
+  fadeLineBrightnessOnEnable: 1
```
Comment (julenka) [inline Assets/MixedRealityToolkit.SDK/Features/UX/Prefabs/Pointers/ParabolicPointer.prefab]: Are these expected?

File: Assets/MixedRealityToolkit.Services/InputSimulation/Editor/InputSimulationService.cs [hunk 10]
```diff
@@ -208,6 +208,8 @@ public bool CheckCapability(MixedRealityCapability capability)
         /// <inheritdoc />
         public override void Initialize()
         {
+            base.Initialize();
```
Comment (julenka) [inline Assets/MixedRealityToolkit.Services/InputSimulation/Editor/InputSimulationService.cs]: Interesting. I wonder if this can break input simulation? Would be good to check that input simulation is still working after adding the base calls.

File: Assets/MixedRealityToolkit.Tests/PlayModeTests/PointerTests.cs [hunk 11]
```diff
@@ -72,6 +75,124 @@ public IEnumerator TestPointerDirectionToCameraDirection()
...
+            interactable.OnClick.AddListener(() => { wasClicked = true; });
+
+            PlayModeTestUtilities.GetInputSimulationService().EnablePointerCache = false;
+
+            var rightHand = new TestHand(Handedness.Right);
+            yield return rightHand.Show(Vector3.right);
+
+            var rightPokePointer = PlayModeTestUtilities.GetPointer<PokePointer>(Handedness.Right);
+            Assert.IsNotNull(rightPokePointer);
+            Assert.IsFalse(rightPokePointer.DestroyOnSourceLost);
```
Comment (julenka) [inline Assets/MixedRealityToolkit.Tests/PlayModeTests/PointerTests.cs]: What is the behavior if `pointer.DestroyOnSourceLost` is true but `PlayModeTestUtilities.GetInputSimulationService().EnablePointerCache` is false? Would a pointer be disabled, but in a cached object list? I would expect the following

When input source is lost....
EnablePointerCache true, DestroySourceLost true => not sure the behavior
EnablePointerCache true, DestroySourceLost false => Pointer gets disabled but moved into the cache
EnablePointerCache false, DestroySourceLost false => Pointer gets disabled
EnablePointerCache false, DestroySourceLost true=> Pointer gets destroyed

File: Assets/MixedRealityToolkit/Interfaces/InputSystem/IMixedRealityPointer.cs [hunk 12]
```diff
@@ -125,5 +125,13 @@ public interface IMixedRealityPointer : IEqualityComparer
         /// Called during the scene query just before the current pointer target changes.
         /// </summary>
         void OnPreCurrentPointerTargetChange();
+
+        /// <summary>
+        /// Resets pointer to initial state. After invoked pointer should be functional and ready for re-use
```
Comment (julenka) [inline Assets/MixedRealityToolkit/Interfaces/InputSystem/IMixedRealityPointer.cs]:
```suggestion
        /// Resets pointer to initial state. After invoked pointer should be functional and ready for re-use.
```

File: Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs [hunk 13]
```diff
@@ -3,6 +3,7 @@
 
 using Microsoft.MixedReality.Toolkit.Utilities;
 using System.Collections.Generic;
+using Unity.Profiling;
```
Comment (julenka) [inline Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs]: Remove?

File: Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs [hunk 14]
```diff
@@ -12,6 +13,32 @@ namespace Microsoft.MixedReality.Toolkit.Input
...
+        /// <summary>
+        /// Control mechanism to enable/disable use of Pointer Cache in request/recycling of pointers by Input System
+        /// </summary>
+        public bool EnablePointerCache
+        {
+            get => enablePointerCache;
+            set
+            {
+                enablePointerCache = value;
+                if (!enablePointerCache)
```
Comment (julenka) [inline Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs]: Do you want to ignore the change if enablePointerCache is already false?

File: Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs [hunk 15]
```diff
@@ -45,13 +72,108 @@ public abstract class BaseInputDeviceManager : BaseDataProvider<IMixedRealityInp
...
+                    DestroyPointerCache();
+                }
+
+                pointerConfigurations = new PointerConfig[initPointerOptions.Length];
+                activePointersToConfig.Clear();
+
+                for (int i = 0; i < initPointerOptions.Length; i++)
+                {
+                    pointerConfigurations[i].profile = initPointerOptions[i];
+                    pointerConfigurations[i].cache = new Stack<IMixedRealityPointer>();
```
Comment (julenka) [inline Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs]: why is it a stack?

Comment (julenka) [reply Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs]: Is this because you're thinking 'access the most recently lost item first' since hands tend to disappear and re-appear?

File: Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs [hunk 16]
```diff
@@ -61,36 +183,168 @@ public abstract class BaseInputDeviceManager : BaseDataProvider<IMixedRealityInp
...
+                    if (option.ControllerType.HasFlag(controllerType) && option.Handedness.HasFlag(controllingHand))
+                    {
+                        IMixedRealityPointer requestedPointer = null;
+
+                        if (EnablePointerCache)
+                        {
+                            var pointerCache = pointerConfigurations[i].cache;
+                            while (pointerCache.Count > 0)
+                            {
+                                var p = pointerCache.Pop();
```
Comment (julenka) [inline Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs]: Why would we ever have more than one pointer of a given controller, handedness in the stack here?

File: Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs [hunk 17]
```diff
@@ -61,36 +183,168 @@ public abstract class BaseInputDeviceManager : BaseDataProvider<IMixedRealityInp
         /// <param name="useSpecificType">Only register pointers with a specific type.</param>
         protected virtual IMixedRealityPointer[] RequestPointers(SupportedControllerType controllerType, Handedness controllingHand)
         {
-            var pointers = new List<IMixedRealityPointer>();
-
-            if ((Service != null) &&
-                (InputSystemProfile != null) &&
-                InputSystemProfile.PointerProfile != null)
+            using (RequestPointersPerfMarker.Auto())
```
Comment (julenka) [inline Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs]: Do you want to leave this in?

File: Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs [hunk 18]
```diff
@@ -61,36 +183,168 @@ public abstract class BaseInputDeviceManager : BaseDataProvider<IMixedRealityInp
...
+                                    pointerComponent.gameObject.SetActive(true);
+
+                                    // We got pointer from cache, continue to next pointer option to review
+                                    requestedPointer = p;
+                                    break;
+                                }
+                            }
+                        }
+
+                        if (requestedPointer == null)
```
Comment (julenka) [inline Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs]: should you only create this pointer if the hand and controller type match the current option? It seems like with this code, even if i have an articulated hand up, you will create all poitners for vr controllers and mouse controllers as well.

File: Assets/MixedRealityToolkit/Providers/UnityInput/MouseDeviceManager.cs [hunk 19]
```diff
@@ -93,22 +91,19 @@ public float WheelSpeed
         /// </summary>
         public MouseController Controller { get; private set; }
 
-        private void ReadProfile()
```
Comment (julenka) [inline Assets/MixedRealityToolkit/Providers/UnityInput/MouseDeviceManager.cs]: Why did you remove this? Will Cursor and wheel speed still be initialized?

File: Assets/MixedRealityToolkit.Services/InputSimulation/Editor/InputSimulationService.cs (hunk 10 above)
Comment (Troy-Ferrell) [reply Assets/MixedRealityToolkit.Services/InputSimulation/Editor/InputSimulationService.cs]: Why would this break input simulation? The hands interaction example scene worked for me in editor which uses this service for the input.

File: Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs (hunk 13 above)
Comment (Troy-Ferrell) [reply Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs]: Why remove? This is used for the RequestPointersPerfMarker. The markers are conditionally removed for release builds
"They are conditionally compiled away and thus have zero overhead in non-Developmenet (Release) builds."

File: Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs (hunk 14 above)
Comment (Troy-Ferrell) [reply Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs]: yea good idea, adding != check

File: Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs (hunk 17 above)
Comment (Troy-Ferrell) [reply Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs]: yes see other comment. Markers are conditionally compiled out for release builds

File: Assets/MixedRealityToolkit/Providers/UnityInput/MouseDeviceManager.cs (hunk 19 above)
Comment (Troy-Ferrell) [reply Assets/MixedRealityToolkit/Providers/UnityInput/MouseDeviceManager.cs]: Not deleted, just moved to end of file. See full file change

File: Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs (hunk 16 above)
Comment (Troy-Ferrell) [reply Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs]: Handedness can equal "any" or both etc. The cache works per pointer option which is mostly 1:1 with a pointer prefab (not a hard expectation). 

image

File: Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs (hunk 15 above)
Comment (Troy-Ferrell) [reply Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs]: yea just easy way to grab earliest item. I don't want a Pointer[] because of re-allocation and copying. I'm pretty sure List<> will also re-allocate and copy for when head is removed.  LinkedList or stack makes it easy to remove item

File: Assets/MixedRealityToolkit.SDK/Features/UX/Prefabs/Pointers/ParabolicPointer.prefab (hunk 9 above)
Comment (Troy-Ferrell) [reply Assets/MixedRealityToolkit.SDK/Features/UX/Prefabs/Pointers/ParabolicPointer.prefab]: Yes these are just the default values in the MRLineRenderer class. At some point they were added but this prefab was never updated. When executed at runtime, Unity just takes the defaults even though there is no entry in the text file. 

Basically no one has opened and saved the prefab asset since these new line renderer fields were added

File: Assets/MixedRealityToolkit.Tests/PlayModeTests/PointerTests.cs (hunk 11 above)
Comment (Troy-Ferrell) [reply Assets/MixedRealityToolkit.Tests/PlayModeTests/PointerTests.cs]: > EnablePointerCache true, DestroySourceLost true => not sure the behavior

If DestroyOnSourceLost is true, the pointer will be destroyed. Whether the pointer is "added" to the cache or not, depends on how quickly the pointer destroys itself since it's not immediate by Unity. 

Technically it's possible for a pointer to be added when it's about to be destroyed is added, or a pointer that is inactive but is destroyed at some future point while in the cache. 

In these cases, the cache scans itself and clears null entries with CleanActivePointers() or with the stack popping while loop scan

File: Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs (hunk 18 above)
Comment (Troy-Ferrell) [reply Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs]: Yes look at line 199
` if (option.ControllerType.HasFlag(controllerType) && option.Handedness.HasFlag(controllingHand))`

We have to pass this gate before even instantiating or accessing the cache for a pointer

**Comment by julenka** [comment]
> @julenka very good points. I think at least the basic test flow at a minimum should be
> 
> * Raise hand
> * Click on objects
> * Hide hand (source lost)
> * Raise hand
> * Click on object at different location
> * Hide hand
> 
> Another potential one (which may already exist...I need to look at pointer events)
> 
> * Raise both hands and focus on same gameobject
> * alternate show/hide of hands (one always remain in focus on button)
> * then hand ray select button

Since this is changing a core part of MRTK I would feel more comfortable approving this change if there were tests included with this PR. I didn't actually see any new tests in the PR, though I did see new utility classes in the tests. Did I miss the file where you added tests to verify that the pointer cache is behaving well? Even with thorough manual testing I feel like this change would benefit from an automated test, since it will at least help us ensure that any future changes to caching won't break stuff.

**Comment by Troy-Ferrell** [comment]
@julenka yes there were 3 new tests added under PointerTests.cs

File: Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs (hunk 17 above)
Comment (keveleigh) [reply Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs]: The docs only mention:

>Methods Begin and End are marked with ConditionalAttribute

Looks like we might be allocating an `Auto` that then does nothing because its internals are compiled out:

>Begin is called in the constructor and End in the Dispose method.

Though it's a struct, so the perf impacts may be minimal in this case

Comment (Troy-Ferrell) [reply Assets/MixedRealityToolkit/Providers/BaseInputDeviceManager.cs]: @keveleigh  True indeed. I know it's every frame, but the two key pieces for me of having minimal impact
1) it's a struct, so on the stack and not on the heap. Thus no GC, basically like allocating some primitives in a function
2) It does not execute begin/end() so computation is does nothing but be allocated on the stack. 

I went with the auto because I like the scoping nature of it. Of course, in the future, probably best practice will be to do the following so get best of both worlds

```csharp
ProfileMarker.Begin();
{
....
}
ProfileMarker.End();
```
